File: dnjs/html.py
from dataclasses import asdict, is_dataclass
from html import escape
import re
from typing import Any, Callable

from dnjs import builtins

SELF_CLOSING = {
    "area",
    "base",
    "br",
    "col",
    "embed",
    "hr",
    "img",
    "input",
    "link",
    "meta",
    "param",
    "source",
    "track",
    "wbr",
}


def make_value_js_friendly(value: builtins.Value) -> builtins.Value:
    if value is None or isinstance(value, (float, int, bool, str, builtins.TrustedHtml)):
        return value
    if isinstance(value, dict):
        return {k: make_value_js_friendly(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [make_value_js_friendly(n) for n in value]
    # we turn functions into null
    if isinstance(value, Callable):
        return None
    if is_dataclass(value):
        return make_value_js_friendly(asdict(value))
    # handle pydantic without importing it
    if hasattr(value, "dict") and callable(value.dict):
        return make_value_js_friendly(value.dict())
    raise RuntimeError(f"unable to make type jsonable {type(value)}")


def to_html(value: builtins.Value, indent: int = 0) -> str:
    assert builtins.is_renderable(value)
    if value is None:
        return ""
    if isinstance(value, builtins.TrustedHtml):
        return ("    " * indent) + value.string
    if isinstance(value, str):
        return ("    " * indent) + escape(value)
    if isinstance(value, (float, int)):
        return ("    " * indent) + str(value)
    # else is vnode
    tag = value["tag"]
    attrs = {**value["attrs"]}
    children = value["children"]

    attrs_str = ""
    for k, v in attrs.items():
        if k == "className":
            k = "class"
            if not v:
                continue
        if v is None or v is False:
            pass
        elif v is True:
            attrs_str += f' {escape(k)}'
        elif isinstance(v, (float, int)):
            attrs_str += f' {escape(k)}="{str(v)}"'
        elif isinstance(v, str):
            attrs_str += f' {escape(k)}="{escape(v)}"'
        else:
            raise RuntimeError(f"unable to convert type {type(v)}")

    is_self_closing = tag in SELF_CLOSING and not children
    html_str = ("    " * indent) + f"<{escape(tag)}{attrs_str}>\n"
    if not is_self_closing:
        if tag in {"pre", "code", "textarea"}:
            html_str = html_str[:-1]  # strip \n
            html_str += "".join(to_html(c, 0) for c in children)
            html_str += f"</{escape(tag)}>\n"
        else:
            html_str += "".join(to_html(c, indent + 1) for c in children) + "\n"
            html_str += ("    " * indent) + f"</{escape(tag)}>\n"
    return html_str

File: dnjs-go/html.go
package main

import (
	"fmt"
	"html"
	"reflect"
	"sort"
	"strings"
)

var selfClosing = []string{
	"area",
	"base",
	"br",
	"col",
	"embed",
	"hr",
	"img",
	"input",
	"link",
	"meta",
	"param",
	"source",
	"track",
	"wbr",
}

func ToHtml(value Value) (string, error) {
	return toHtml(value, 0)
}

func toHtml(value Value, indent int) (string, error) {
	indentString := strings.Repeat("    ", indent)
	if !isRenderable(value) {
		return "", fmt.Errorf("value cannot be converted to html: " + fmt.Sprint(value))
	}
	if value == nil {
		return "", nil
	}
	if reflect.TypeOf(value) == reflect.TypeOf(TrustedHtml{}) {
		return indentString + value.(TrustedHtml).Html, nil
	}
	switch v := value.(type) {
	case string:
		return indentString + html.EscapeString(v), nil
	case int64:
		return indentString + fmt.Sprint(v), nil
	case float64:
		return indentString + fmt.Sprint(v), nil
	}
	// else is vnode
	vNode := value.(map[string]Value)
	_, hasTag := vNode["tag"]
	_, hasAttrs := vNode["attrs"]
	_, hasChildren := vNode["children"]
	if !hasTag || !hasAttrs || !hasChildren {
		return "", fmt.Errorf("value must have tag, attrs, children attributes")
	}

	tag, ok := vNode["tag"].(string)
	if !ok {
		return "", fmt.Errorf("tag must be a string")
	}
	attrs, ok := vNode["attrs"].(map[string]Value)
	if !ok {
		return "", fmt.Errorf("attributes must be a map of string to value")
	}
	children, ok := vNode["children"].([]Value)
	if !ok {
		return "", fmt.Errorf("children must be an array of values")
	}

	// sort keys first
	keys := []string{}
	for k, _ := range attrs {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	attrsStr := ""
	for _, k := range keys {
		v := attrs[k]
		if k == "className" {
			k = "class"
			if reflect.ValueOf(v).IsZero() {
				continue
			}
		}
		if v == nil || v == false {
			// pass
		} else if v == true {
			attrsStr += " " + html.EscapeString(k)
		} else {
			switch v_ := v.(type) {
			case string:
				attrsStr += " " + html.EscapeString(k) + "=\"" + html.EscapeString(v_) + "\""
			case int64:
				attrsStr += " " + html.EscapeString(k) + "=\"" + fmt.Sprint(v_) + "\""
			case float64:
				attrsStr += " " + html.EscapeString(k) + "=\"" + fmt.Sprint(v_) + "\""
			default:
				return "", fmt.Errorf("unable to render attribute type")
			}
		}
	}
	isSelfClosing := containsString(selfClosing, tag) && len(children) == 0
	htmlStr := indentString + "<" + html.EscapeString(tag) + attrsStr + ">\n"
	if !isSelfClosing {
		if containsString([]string{"pre", "code", "textarea"}, tag) {
			htmlStr = string(([]rune(htmlStr))[0 : len([]rune(htmlStr))-1]) // strip \n
			for _, c := range children {
				innerHtmlStr, err := toHtml(c, 0)
				if err != nil {
					return "", err
				}
				htmlStr += innerHtmlStr
			}
			htmlStr += "</" + html.EscapeString(tag) + ">\n"
		} else {
			for _, c := range children {
				innerHtmlStr, err := toHtml(c, indent+1)
				if err != nil {
					return "", err
				}
				htmlStr += innerHtmlStr
			}
			htmlStr += "\n" + indentString + "</" + html.EscapeString(tag) + ">\n"
		}
	}
	return htmlStr, nil
}
